### PR TITLE
Initialise the CLI with ink powered by Commander

### LINKED ISSUE #2: Create watch CLI command
Create `watch` command.

### PR DESCRIPTION
**Pull request by saphal1998**
## Description

- Used Commander to define the commands in the CLI
- Used Ink to render the command

As of now, only the watch command is working, however, we can follow this directory structure to work on the rest of the application.

## Testing Plan

As of now, you can only check if a test is present or not in `jslings`

Screenshot 2020-09-24 at 4 18 51 PM

### CODE REVIEW

File: src/cli/index.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    console.log("  $ jslings watch");
+    console.log("  $ jslings hint");
+  });
+
+program
+  .command("watch")
+  .alias("w")
+  .description("jslings interactive code testing UI")
+  .option("-t, --test <testName>", "path to the test")
+  .action(async ({ test }: Command) => {
```
Comment (amorriscode) [inline src/cli/index.ts]: We should check if they passed in `test`, if not, find out which one they are on.

File: src/cli/ui/ui.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import React from "react";
```
Comment (amorriscode) [inline src/cli/ui/ui.tsx]: Doesn't necessarily need to be resolved but I'm more used to seeing `/components` instead of `/ui`.

File: package.json [hunk 3]
```diff
@@ -1,14 +1,69 @@
...
+    "ink-testing-library": "^2.0.1",
+    "jest-junit": "^11.1.0",
+    "nodemon": "^2.0.4",
+    "ts-jest": "^26.4.0",
+    "ts-node": "^9.0.0",
+    "typescript": "^4.0.3",
+    "xo": "^0.33.1"
+  },
+  "scripts": {
+    "cli": "ts-node ./src/cli/index.ts",
```
Comment (amorriscode) [inline package.json]: Could use `jslings` so they can `yarn jslings...` instead of the generic `cli`?

**Review by amorriscode — CHANGES_REQUESTED** [review]
Awesome work! Really minor changes requested.

**Comment by amorriscode** [comment]
Glad we get to use `ink`! 🐙

File: src/cli/index.ts (hunk 1 above)
Comment (saphal1998) [reply src/cli/index.ts]: Do you mean that if not provided the -t flag, just watch the current test to complete? 

I'm thinking we can have a `--all` flag to see the tests available, which goes in `utils`.

File: src/cli/ui/ui.tsx (hunk 2 above)
Comment (saphal1998) [reply src/cli/ui/ui.tsx]: Planning to break this into other components, i.e. the UI folder can have a folder `components` and then we can import using 

```ts
import {Component} from './component/Component'
``` 

Thoughts?

File: src/cli/index.ts (hunk 1 above)
Comment (amorriscode) [reply src/cli/index.ts]: Yeah. For example `rustlings` doesn't require you to pass the test. If you `rustlings watch`, you just pick up where you left off. I think it'd be great if we mimic that functionality! Makes it easier for the user to keep going. 🏃

File: package.json (hunk 3 above)
Comment (saphal1998) [reply package.json]: Yup, this seems fine, making this change

**Comment by saphal1998** [comment]
> Glad we get to use `ink`! 🐙

Well, this is still my first time building a UI with `ink`. Hopefully it works out 🤞🏼

File: src/cli/index.ts (hunk 1 above)
Comment (saphal1998) [reply src/cli/index.ts]: I see, yeah I think this will go in `utils` tbh, will try this out

File: src/cli/ui/ui.tsx (hunk 2 above)
Comment (amorriscode) [reply src/cli/ui/ui.tsx]: What else will go under `/ui`? I guess I just don't see the benefit of having a `/ui` folder if it doesn't contain anything but another folder! In the end it's not the end of the world. :)

File: src/cli/index.ts (hunk 1 above)
Comment (amorriscode) [reply src/cli/index.ts]: @saphal1998 we can just create an issue for this and do it in a different PR!

Comment (saphal1998) [reply src/cli/index.ts]: Yeah, I was just brainstorming this, and it seems like we would need to add a bunch of stuff to do this. Will make a seperate PR for this

File: src/cli/ui/ui.tsx (hunk 2 above)
Comment (saphal1998) [reply src/cli/ui/ui.tsx]: Yeah fair, I was hoping to get a test file viewer in, but that is probably a stretch goal too, for now this should be fine. Making this change

File: README.md [hunk 4]
```diff
@@ -1 +1,3 @@
-# jslings
\ No newline at end of file
+# jslings
+
+`jslings` can help YOU master Javascript on the terminal
```
Comment (XifeiNi) [inline README.md]: Quick comment: Maybe you should revert back README for now. 
This file deserves a separate PR to compose, and we are not there yet.

**Review by amorriscode — APPROVED** [review]

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,12 @@
+root = true
+
+[*]
+indent_style = tab
```
Comment (amorriscode) [inline .editorconfig]: Maybe we can use soft spaces instead?

```suggestion
indent_style = space
indent_size = 4
```
I noticed some spacing inconsistencies.

**Review by amorriscode — CHANGES_REQUESTED** [review]
Just a suggestion based on spacing inconsistencies I originally missed.

**Review by amorriscode — APPROVED** [review]

File: .gitignore [hunk 6]
```diff
@@ -1 +1,323 @@
...
+# Windows Installer files
+*.cab
+*.msi
+*.msix
+*.msm
+*.msp
+
+# Windows shortcuts
+*.lnk
+
```
Comment (XifeiNi) [inline .gitignore:322]:
```suggestion

# editor configuration file 
.editorconfig
```

File: .editorconfig [hunk 7]
```diff
@@ -0,0 +1,13 @@
+root = true
```
Comment (XifeiNi) [inline .editorconfig:1]: This is your own editor configuration right? It probably should not in the repo. Delete this and include this file in .gitignore?

File: .prettierrc [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (XifeiNi) [inline .prettierrc:1]: This is also your local configuration for style checking right? Delete and include this in gitignore

File: src/cli/components/landingUI.tsx [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                </Text>
+            );
+        } else if (fileStatus === FileStatus.INCONCLUSIVE) {
+            return (
+                <Text>
+                    <Text color="red">{testFileName}</Text> is inconclusive,
+                    please provide a uniquely identifiable test name
+                </Text>
+            );
+        }
```
Comment (XifeiNi) [inline src/cli/components/landingUI.tsx]: refactor this into a function called `renderByFileStatus`?

File: src/cli/components/landingUI.tsx [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    return (
+        <Text>
+            Invalid test <Text color="red">{testFileName}</Text> provided
+        </Text>
+    );
```
Comment (XifeiNi) [inline src/cli/components/landingUI.tsx:37]:
```suggestion
    } else {
            return (
                   <Text>
                          Invalid test <Text color="red">{testFileName}</Text> provided
                   </Text>
             );
    }
```
Add an else block to make this logic more clear

File: src/cli/index.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env node
+import React from "react";
```
Comment (XifeiNi) [inline src/cli/index.ts:3]:
```suggestion

import React from "react";
```
Blank line between usr/bin instruction and ts import?

File: src/cli/utils/fileSystem.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        }
+        // Should never reach here
+        return FileStatus.FAILURE;
+    } catch (_) {
+        return FileStatus.FAILURE;
+    }
+};
+
+/**
+ * A utility function to list all the files present inside a folder
```
Comment (XifeiNi) [inline src/cli/utils/fileSystem.ts]:
```suggestion
 * A utility function to list all the file names present inside a folder
```
bc you returned strings rather than file node

File: src/cli/utils/fileSystem.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            .readdirSync(path)
+            .map((file) => file.startsWith(partialFileName));
+        if (files.length > 0) {
+            if (files.length === 1) {
+                return FileStatus.SUCCESS;
+            }
+            return FileStatus.INCONCLUSIVE;
+        }
+        // Should never reach here
+        return FileStatus.FAILURE;
```
Comment (XifeiNi) [inline src/cli/utils/fileSystem.ts:24]:
```suggestion
            if (files.length === 1) {
                return FileStatus.SUCCESS;
            } else {
                return FileStatus.INCONCLUSIVE;
            }
```

File: src/types/exercises.types.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+export enum FileStatus {
+    INCONCLUSIVE = "X", // Inconclusive simply means we could not uniquely identify the filename in our directory structure. A drawback of using partial test names
+    SUCCESS = "S",
+    FAILURE = "F",
+}
+
+/**
+ * Each exercise should follow this interface and contain these dataPoints
+ */
+interface Exercise {
```
Comment (XifeiNi) [inline src/types/exercises.types.ts:22]: should this be exported too?

File: src/cli/components/landingUI.tsx [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        if (fileStatus === FileStatus.SUCCESS) {
+            return (
+                <Text>
+                    Hello <Text color="green">{testFileName}</Text>
+                </Text>
+            );
+        } else if (fileStatus === FileStatus.INCONCLUSIVE) {
+            return (
+                <Text>
+                    <Text color="red">{testFileName}</Text> is inconclusive,
```
Comment (XifeiNi) [inline src/cli/components/landingUI.tsx]:
```suggestion
                    There exists more than one files with the name <Text color="red">{testFileName}</Text>
```
This might be clearer than "inconclusive"

**Review by XifeiNi — CHANGES_REQUESTED** [review]
😃

File: .gitignore (hunk 6 above)
Comment (amorriscode) [reply .gitignore:322]: The `.editorconfig` should be in source control so that everyone is consistently styling the files.

File: .prettierrc (hunk 8 above)
Comment (amorriscode) [reply .prettierrc:1]: This should actually be checked into source control. It ensures everyone's files are consistently styled. 😄

File: .editorconfig (hunk 7 above)
Comment (amorriscode) [reply .editorconfig:1]: The `.editorconfig` should be in source control so that everyone is consistently styling the files.

File: src/cli/components/landingUI.tsx (hunk 9 above)
Comment (saphal1998) [reply src/cli/components/landingUI.tsx]: Yeah, well, this is the core UI I just wanted something functional at this point. I may even be removing this, this was added just for demonstration

File: src/cli/components/landingUI.tsx (hunk 10 above)
Comment (saphal1998) [reply src/cli/components/landingUI.tsx:37]: The reason I chose to omit `else` is get flatter code. Writing an `else`, just adds extra indentation for no reason. Thoughts?

File: src/cli/utils/fileSystem.ts (hunk 12 above)
Comment (saphal1998) [reply src/cli/utils/fileSystem.ts]: Good catch!

File: src/cli/index.ts (hunk 11 above)
Comment (saphal1998) [reply src/cli/index.ts:3]: Nice catch!

File: src/cli/utils/fileSystem.ts (hunk 13 above)
Comment (saphal1998) [reply src/cli/utils/fileSystem.ts:24]: I think this is incorrect, because `INCONCLUSIVE` differs from `FAILURE`. Failure implies that the file couldn't be found, however inconclusive implies file could not be uniquely found.

File: src/types/exercises.types.ts (hunk 14 above)
Comment (saphal1998) [reply src/types/exercises.types.ts:22]: It should be in the future, didn't require to export this as of now, so I didn't export it

File: src/cli/components/landingUI.tsx (hunk 15 above)
Comment (saphal1998) [reply src/cli/components/landingUI.tsx]: Nice catch!

File: src/cli/components/landingUI.tsx (hunk 10 above)
Comment (XifeiNi) [reply src/cli/components/landingUI.tsx:37]: Semantic clarity is more important in my thoughts

File: src/cli/utils/fileSystem.ts (hunk 13 above)
Comment (XifeiNi) [reply src/cli/utils/fileSystem.ts:24]: Yes, but you will never reach FAILURE in the chunk of code so what's the point of having it anyway

File: src/cli/components/landingUI.tsx (hunk 9 above)
Comment (XifeiNi) [reply src/cli/components/landingUI.tsx]: So rustling basically name files with its process ID and hope it eventually be unique. I think maybe leave a comments on the code to show that this might be removed. 

Because it will probably get removed, it's more of a reason to refactor this into another function so to allow the ease of future changes (change to different UI or remove it completely, etc)

File: .prettierrc (hunk 8 above)
Comment (XifeiNi) [reply .prettierrc:1]: @amorriscode Considering a project to be open sourced, you might get pull requests by very different kinds of people. Right now there's only 4 ppl, and it's really easy to request everyone uses the same formatter. 
But for an open source proj that actually has some impact on the community, keeping a formatter in repo isn't actually a legitimate practice.

File: .editorconfig (hunk 7 above)
Comment (XifeiNi) [reply .editorconfig:1]: ditto as below @amorriscode

**Comment by XifeiNi** [comment]
A lot of future PR will be based off this one, so I would really appreciate if we can get this merged soon. @victorphoenix3 do you have any review comments?

File: .prettierrc (hunk 8 above)
Comment (amorriscode) [reply .prettierrc:1]: I actually disagree with you there. It's even _more_ important when it is open source and has impact on the community. Having a project where the formatting is different in every single file would be a nightmare. That's why things like Prettier exist in the first place!

In Rust and Go, for example, everything is build into the language (`rustfmt`, `go fmt`). Unfortunately we aren't blessed with something so convenient so each project is on their own to deal with it.

As another example, [you can see the `.prettierrc.js` file for React](https://github.com/facebook/react/blob/master/.prettierrc.js). If they do it, I'm sure we can! 😄

File: .gitignore (hunk 6 above)
Comment (amorriscode) [reply .gitignore:322]: Example: [React keeping `.editorconfig` in source control](https://github.com/facebook/react/blob/master/.editorconfig).

**Review by victorphoenix3 — APPROVED** [review]
